File: Leetcode/Dynamic/5_longest-palindromic-substring/main.cpp
// 5. 最长回文子串
// Created by bob on 2022/4/19.
//
#include<string>
#include<vector>

using namespace std;

class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        int maxlen = 1;
        int begin = 0;
        vector<vector<int>> dp(n, vector<int>(n));
        for (int i = 0; i < n; ++i) {
            dp[i][i] = true;
        }
        // 先枚举子串长度
        for (int L = 2; L <= n; ++L) {
            // 枚举左边界
            for (int i = 0; i < n; ++i) {
                int j = L + i - 1; // 右边界
                if (j >= n) {
                    break;
                }
                if (s[i] != s[j]) {
                    dp[i][j] = false;
                } else {
                    if (j - i < 3) {    // 最小情况
                        dp[i][j] = true;
                    } else {
                        dp[i][j] = dp[i + 1][j - 1];
                    }
                }
                // 统计最大长度
                if (dp[i][j] && j - i + 1 > maxlen) {
                    maxlen = j - i + 1;
                    begin = i;
                }

            }

        }
        return s.substr(begin, maxlen);
    }
};

int main() {
    string s = "babad";
    string res = Solution().longestPalindrome(s);
    printf("%s", res.c_str());     // expect: bab
}

File: Leetcode/Dynamic/5_longest-palindromic-substring/main.py


class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        dp = [[False] * n for _ in range(n)]
        for i in range(n):
            dp[i][i] = True

        max_len = 1
        start = 0
        for l in range(2, n+1):
            for i in range(0, n):
                # j - i + 1 = L
                j = l + i - 1
                if j >=n:
                    break
                if s[i] != s[j]:
                    dp[i][j] = False
                else:
                    if j - i < 3:
                        dp[i][j] = True
                    else:
                        dp[i][j] = dp[i+1][j-1]

                if dp[i][j] and l > max_len:
                    max_len = l
                    start = i

        return s[start:start + max_len]

if __name__ == '__main__':
    s = "babad"
    res = Solution().longestPalindrome(s)
    print(res)